llvm/include/llvm/MC/MCStreamer.h

```diff
@@ -1003,6 +1003,7 @@ class MCStreamer {
   virtual void emitCFIRegister(int64_t Register1, int64_t Register2);
   virtual void emitCFIWindowSave();
   virtual void emitCFINegateRAState();
+  virtual void emitCFICompactUnwindEncoding(unsigned Encoding);
```

Comment:
A lot of these emitCFI* virtuals are overridden in MCAsmStreamer.cpp.  Does that make sense here?  I'm not familiar enough with these yet, so I think I'm better off asking you than waiting until I can learn it all.

Reply:
The unwinding encoding is inherently specific to the Apple Mach-O file format and a processor architecture target. Any value for `Encoding` parameter besides 0 (ie. letting LLVM use the architecture specific conversation which uses the complex layering and virtual overrides) would have to be generated in the NativeAOT ILCompiler with the knowledge of specific target (ie. the values have different meaning on x64 and Arm64).

There are certain function prolog patterns that can be expressed by compact unwinding but which LLVM cannot infer automatically. For example, when there is more than one page of stack reserved space the JIT may generate a `__chkstk` call to probe the stack pages. The compact unwinding code in that case is "RBP-based frame with indirect stack size" type. This unwind code works by encoding the offset of "<value>" in the "sub rsp, <value>" instruction to determine the frame size. However, since the `__chkstk` pattern uses different instructions LLVM doesn't recognize it. The ILCompiler/JIT can recognize it and compose the value of the unwind code by itself because it knows these patterns. Obviously the code would have to be highly specific for both the osx-x64 and osx-arm64 targets. In the end it is an optimization that may or may not be worth the effort. However, since we know what the JIT emits it's much easier to implement in ILCompiler than by LLVM trying to second guess it (and guessing it wrong leads to runtime crashes that are hard to debug).

Reply:
Here's an example of prolog that LLVM cannot handle by itself:
```
0000000000000000 55                              push      rbp
0000000000000001 4157                            push      r15
0000000000000003 4156                            push      r14
0000000000000005 4155                            push      r13
0000000000000007 4154                            push      r12
0000000000000009 53                              push      rbx
000000000000000A 4C8D9C2468EFFFFF                lea       r11,[rsp-1098h]
0000000000000012 E800000000                      call      0000000000000017h // __chkstk
0000000000000017 498BE3                          mov       rsp,r11
```

On second inspection I see that this would need a codegen change to get positive offset in the code to generate valid compact encoding. The `<value>` in `lea       r11,[rsp-1098h]` is negative offset which cannot be encoded.

Reply:
Sorry, my question was obviously very unclear (though separately it was still useful to see your explanation and example).

This diff adds a new virtual function to `MCStreamer`.  `MCAsmStreamer` is a derived class of `MCStreamer` used for writing textual assembly files.  The other `emitCFIxxx` functions have overrides in `MCAsmStreamer`, which suggests that this function also should affect the asm output.  I don't know if that means an `emitCFICompactUnwindEncoding` override that simply adds a directive or if something more complicated would need to be done.  If the correct way isn't an override, then perhaps `emitCFICompactUnwindEncoding` doesn't need to be virtual.

Reply:
My bad :-) I wanted to explain that but got side-tracked with my line of thinking...

You are right, the `emitCFICompactUnwindEncoding` method doesn't have to be virtual. In fact, the only reason it needs to exist is that `getCurrentDwarfFrameInfo` is private and `getDwarfFrameInfos` returns `const` reference, so neither of those existing methods can be used by objwriter directly.